dietpi/dietpi-software

```diff
@@ -6547,6 +6560,21 @@ Mycroft AI requires at least 2 GiB memory on first start. We will now increase y
 
 		fi
 
+		software_id=177 # TasmoAdmin
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == 1 )); then
+
+			Banner_Installing
+			DEPS_LIST="$PHP_NAME-curl $PHP_NAME-zip" # https://github.com/reloxx13/TasmoAdmin#linux
+
+			Download_Install 'https://github.com/reloxx13/TasmoAdmin/archive/master.zip'
+			cp -a TasmoAdmin-master/tasmoadmin/ /var/www/tasmoadmin/
+			rm -R TasmoAdmin-master
```

Comment:
@svh1985 
What you thin about `mv TasmoAdmin-master/tasmoadmin /var/www/tasmoadmin`?

Reply:
Sounds good! I looked at other implementations and copied that :P

Reply:
@svh1985 
Ah what I did not think about:
- `mv TasmoAdmin-master/tasmoadmin /var/www/tasmoadmin` will fail if the target dir already exists, thus on reinstalls it will not be touched.
- `cp -a TasmoAdmin-master/tasmoadmin/ /var/www/tasmoadmin/` merges the dir inside, although currently not sure if this leads to `/var/www/tasmoadmin/tasmoadmin` if target dir exists.

Actually to assure clean reinstalls:
- Does TasmoAdmin have an internal (web UI or CLI-based) updater? In this case we should simply skip downloading and installing `/var/www/tasmoadmin` if this dir already exists. So reinstalls will re-apply configs, install pre-reqs, PHP and webserver modules, but the PHP scripts themselves stay untouched for internal updater, which e.g. might include some required migration steps and such.
- If no internal updater exist, I guess `/var/www/tasmoadmin` contains some config file or other files/dirs that are expected to be changed/customised by user? We could then merge those files from existing dir to fresh download, remove target dir and move fresh download with custom file/dir merged into place.
- If all custom/settings are saved outside of `/var/www/tasmoadmin`, we could completely remove existing dir and move the new in place.

Reply:
> @svh1985
> * Does TasmoAdmin have an internal (web UI or CLI-based) updater? In this case we should simply skip downloading and installing `/var/www/tasmoadmin` if this dir already exists. So reinstalls will re-apply configs, install pre-reqs, PHP and webserver modules, but the PHP scripts themselves stay untouched for internal updater, which e.g. might include some required migration steps and such.

There is a way to update via the web page itself, I can check if we can also use that method from the command line.
![SelfUpdate_-_TasmoAdmin](https://user-images.githubusercontent.com/5213210/66331540-1cf60580-e933-11e9-942d-ecb675597baa.png)

> * If no internal updater exist, I guess `/var/www/tasmoadmin` contains some config file or other files/dirs that are expected to be changed/customised by user? We could then merge those files from existing dir to fresh download, remove target dir and move fresh download with custom file/dir merged into place.


> * If all custom/settings are saved outside of `/var/www/tasmoadmin`, we could completely remove existing dir and move the new in place.

All relevant data is stored in : `/var/www/tasmoadmin/data`, so I think this should be backuped somehow.

Reply:
@svh1985
In this case, IMO it is the best solution to skip download if `/var/www/tasmoadmin` exist and instead prompt an info that user should use the internal updater. All other things like pre-reqs and required webserver drop-in configs should still be applied, so it is only about the web app scripts themselves.

Check how we do this with Pydio as an example: https://github.com/MichaIng/DietPi/blob/master/dietpi/dietpi-software#L5078-L5092

This is simply safest, e.g. if a newer version does not work with old configs, e.g. some migration steps required that are done by internal updater then. Also if some database is used, this is often the case.

If there is some CLI available to call the internal updater, this would be an alternative, but depends as well on if e.g. some (optionally) interactive steps are done that would need to be automated on non-interactive installs, respectively this software title then excluded for non-interactive dietpi-software calls.

Good to know about the config/data dir regardless. Long term I'd like to have some software backup feature that creates backups for certain software configs and data, also to allow easier migration of systems to new SBCs/images.

Reply:
I've added that part, but for some reason the name of the software is not shown:
![DietPi_VirtualBox](https://user-images.githubusercontent.com/5213210/66414560-f0f18780-e9f9-11e9-8138-e64562d29542.png)

```
G_DIETPI-NOTIFY 2 "**${aSOFTWARE_WHIP_NAME[$software_id]}** install dir \"/var/www/pydio\" already exists. Download and install steps will be skipped.
 - Please manually backup your config files+data, remove the install dir and rerun \"dietpi-software (re)install $software_id\" if you need to reinstall.
 - If you want to update the **${aSOFTWARE_WHIP_NAME[$software_id]}** instance, please use the internal updater from WebUI."
```

Reply:
@svh1985
Ah this is due to recent change, I renamed the `aSOFTWARE_WHIP_NAME` to `aSOFTWARE_NAME`. Is the first still used/referenced somewhere in current code? At least the wiki has already been updated.

Reply:
Okay, thanks for checking 😄 I'll change it.